contracts/products/SoteriaCoverageProduct.sol

```diff
@@ -0,0 +1,503 @@
+        uint256 coverAmount = activeCoverAmount();
+        coverAmount = coverAmount + newCoverAmount_ - currentCoverAmount;
+        _coverAmountOf[policyID] = newCoverAmount_;
+        _activeCoverAmount = coverAmount;
+        _updatePolicyManager(coverAmount);
+        emit PolicyUpdated(policyID);
+    }
+
+    /**
+     * @notice Deposits funds for policy holders.
+     * @param policyholder_ The holder of the policy.
+    */
+    function deposit(address policyholder_) external payable override whileUnpaused {
+        _deposit(policyholder_, msg.value);
+    }
+
+    /**
+     * @notice Charge premiums for each policy holder.
+     * @param holders_ The policy holders.
+     * @param premiums_ The premium amounts in `wei` per policy holder.
+    */
+    function chargePremiums(address[] calldata holders_, uint256[] calldata premiums_) external payable override onlyGovernance whileUnpaused {
+        uint256 count = holders_.length;
+        require(count == premiums_.length, "length mismatch");
+        require(count <= policyCount(), "policy count exceeded");
+        uint256 amountToPayTreasury = 0;
+
+        for (uint256 i = 0; i < count; i++) {
+            // skip computation if policy inactive (coverAmount == 0)
+            if (coverAmountOf(policyByOwner(holders_[i])) == 0) {
+                continue;
+            }
+
+            if (_funds[holders_[i]] >= premiums_[i]) {
+                amountToPayTreasury += premiums_[i];
+                _funds[holders_[i]] -= premiums_[i];
+                emit PremiumCharged(holders_[i], premiums_[i]);
+            } else {
+                uint256 partialPremium = _funds[holders_[i]];
+                amountToPayTreasury += partialPremium;
+                _funds[holders_[i]] = 0;
+                // turn off the policy
+                _closePolicy(holders_[i]);
+                emit PremiumPartiallyCharged(holders_[i], premiums_[i], partialPremium);
+            }  
+        }
+        // transfer premium to the treasury
+        ITreasury(payable(_registry.treasury())).routePremiums{value: amountToPayTreasury}();
+    }
+
+    /**
+     * @notice Cancel and burn a policy.
+     * User will receive their deposited funds.
+     * Can only be called by the policyholder.
+     * @param policyID_ The ID of the policy.
+     */
+     function cancelPolicy(uint256 policyID_) external override nonReentrant {
+        require(_exists(policyID_), "invalid policy");
+        require(ownerOfPolicy(policyID_) == msg.sender, "!policyholder");
+        uint256 refundAmount = funds(msg.sender);
+        _activeCoverAmount -= coverAmountOf(policyID_);
+        _coverAmountOf[policyID_] = 0;
+
+        // send deposited fund to the policyholder
+        if (refundAmount > 0) Address.sendValue(payable(msg.sender), refundAmount);
+        _funds[msg.sender] = 0;
+        _updatePolicyManager(_activeCoverAmount);
+        emit PolicyCanceled(policyID_);
+    }
+
```

Comment:
Should we add a function for user to withdraw funds? I can imagine a situation where the user originally deposited 10 ETH, only needs 0.05 ETH to meet this epoch's premium and wants to withdraw 5 ETH to do something else. Currently to get the funds back, the user has to call cancelPolicy(), or wait until chargePremium() executes the branch that cancels their account. The following user flow would be annoying just to withdraw some funds: "cancel the entire policy => re-activate policy". Below is a prototype for such a function

```
    /**
     * @notice Withdraw ETH funds
     * @param amount Amount policyholder desires to withdraw
     * Question - should governance be able to withdraw funds for a user?
     * Note - User has to be aware that if they withdraw funds such that they cannot meet the next premium, their account will be cancelled
     */
    withdrawFunds(uint256 amount_) external override nonReentrant whileUnpaused {
      require(_funds[msg.sender] > 0, "You have no funds");
      require(amount_ > 0, "Error - zero amount");
      require(_funds[msg.sender] >= amount_, "You cannot withdraw this amount");
      _funds[msg.sender] -= amount_;
      Address.sendValue(payable(msg.sender), amount_);
      emit FundsWithdrawn(msg.sender, amount_)
    }
```